library/include/ck/library/tensor_operation_instance/gpu/batched_gemm_softmax_gemm_permute.hpp

```diff
@@ -100,85 +185,67 @@ void add_device_batched_gemm_softmax_gemm_permute_xdl_cshuffle_bf16_bf16_bf16_bf
                                                             PassThrough,
                                                             MaskingSpecialization::MaskDisabled>>>&
         instances);
-
-template <typename ADataType,
+template <index_t NumDimG,
```

Comment:
I think the mechanism is generally good and it provides a unified interface that  
1. provides in lib instances by calling
overloads of `add_device_batched_gemm_mutiple_d_softmax_gemm_permute_xdl_cshuffle_gmk_gnk_gno_gmo_instances`  declared previously in this file, and defined in various translation units.
2. provides out of lib instances by letting the users to decl and define `add_device_batched_gemm_mutiple_d_softmax_gemm_permute_xdl_cshuffle_gmk_gnk_gno_gmo_instances` overloads for their perticular needs

But for the second case, I think ck developer should **document here for users/devs**, that to do so, he should:
1. **forward decl** user overloads for `add_device....` before including this header
2. provide definition `add_device....` somewhere.

Otherwise, common users might not know how to use. for "forward decl user overloads for `add_device....` before including this header", might be a issue, and I am not sure if it worth solving by changing it as follows:
```cpp
// a templated prototype
template<typename T>
void add_device_batched_gemm_mutiple_d_softmax_gemm_permute_xdl_cshuffle_gmk_gnk_gno_gmo_instances(T& instances);

template<>
void add_device_batched_gemm_mutiple_d_softmax_gemm_permute_xdl_cshuffle_gmk_gnk_gno_gmo_instances(...) {
// specialization instead of overloads
}

so that the user can provide decl after the inclusion instead of before, thus, in a more natural way.
```